client/lib/track-element-size/index.ts

```diff
@@ -0,0 +1,109 @@
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @param callback The function to call back on changes. Takes a single parameter: `boundingClientRect`.
+ *
+ * @returns The ref to be set on the consumer component.
+ */
+export function useWindowResizeCallback(
+	callback: ( boundingClientRect: UseWindowResizeRect ) => any
+) {
+	const lastRect = useRef< UseWindowResizeRect | null >( null );
+	const [ element, setElement ] = useState< Element | null >( null );
+
+	const callbackRef: UseWindowResizeCallback = useCallback( node => {
+		if ( node ) {
+			setElement( node );
+		}
+	}, [] );
+
+	useEffect( () => {
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = ( rect: UseWindowResizeRect ) => {
+			if ( rect ) {
+				// Avoid notifying consumer if nothing's changed.
+				if (
+					lastRect.current === null ||
+					lastRect.current.bottom !== rect.bottom ||
+					lastRect.current.left !== rect.left ||
+					lastRect.current.right !== rect.right ||
+					lastRect.current.top !== rect.top ||
+					lastRect.current.width !== rect.width ||
+					lastRect.current.height !== rect.height
+				) {
+					lastRect.current = rect;
+					return callback( rect );
+				}
+			}
+		};
+
+		// Measure the element in the DOM.
+		// Uses `afterLayoutFlush` to avoid causing layout thrashing.
+		const measureElement = afterLayoutFlush( () => {
+			if ( element && element.getBoundingClientRect ) {
+				handleRectChange( element.getBoundingClientRect() );
+			}
+		} );
+		const throttledMeasureElement = throttle( measureElement, THROTTLE_RATE );
+
+		// Set up subscription.
+		if ( element && callback ) {
+			// Measure element so that the callback is invoked at least once, even if
+			// there are no window resize events.
+			throttledMeasureElement();
+			window.addEventListener( 'resize', throttledMeasureElement );
+
+			// Unsubscribe.
+			return () => {
+				window.removeEventListener( 'resize', throttledMeasureElement );
+				measureElement.cancel();
+				throttledMeasureElement.cancel();
+			};
+		}
+	}, [ element, callback ] );
+
+	return callbackRef;
+}
+
+/**
+ * React hook that subscribes a consumer to changes to the bounding client rect of an element, based
+ * on window resize events.
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @returns A tuple with the ref to be set on the consumer component, and the current rect.
+ */
+export function useWindowResizeRect(): [UseWindowResizeCallback, UseWindowResizeRect | null] {
+	const [ rect, setRect ] = useState< UseWindowResizeRect | null >( null );
+
+	const callback = useCallback(
+		( boundingClientRect: UseWindowResizeRect ) => setRect( boundingClientRect ),
+		[]
+	);
+	const callbackRef = useWindowResizeCallback( callback );
+
+	return [ callbackRef, rect ];
```

Comment:
`setRect` from `useState` is guaranteed to be the same identical function on every call. The `useCallback` memoization is not neccessary for this native low-level function. The whole hook can be this simple:
```js
function useWindowResizeRect(): [ MutableRefObject< Element | null >, ClientRect | null ] {
  const [ rect, setRect ] = useState< ClientRect | null >( null );
  const callbackRef = useWindowResizeCallback( setRect );
  return [ callbackRef, rect ];
}
```

Reply:
Right, since our callback is just to call `setRect` with the same parameters, we can just use `setRect` directly instead of having a memoised arrow function that calls it 👍

Reply:
Thank you for making these changes in your commit! 👍